packages/@aws-cdk/aws-stepfunctions-tasks/lib/ecs/run-task.ts

```diff
@@ -0,0 +1,376 @@
+      return Array.prototype.concat([], ...xs);
+    }
+
+    function noEmpty<A>(xs: A[]): A[] | undefined {
+      if (xs.length === 0) {
+        return undefined;
+      }
+      return xs;
+    }
+  }
+}
+
+/**
+ * Run a Task on ECS or Fargate
+ */
+export class EcsRunTask extends sfn.TaskStateBase implements ec2.IConnectable {
+  private static readonly SUPPORTED_INTEGRATION_PATTERNS: sfn.IntegrationPattern[] = [
+    sfn.IntegrationPattern.REQUEST_RESPONSE,
+    sfn.IntegrationPattern.RUN_JOB,
+    sfn.IntegrationPattern.WAIT_FOR_TASK_TOKEN,
+  ];
+
+  /**
+   * Manage allowed network traffic for this service
+   */
+  public readonly connections: ec2.Connections = new ec2.Connections();
+
+  protected readonly taskMetrics?: sfn.TaskMetricsConfig;
+  protected readonly taskPolicies?: iam.PolicyStatement[];
+
+  private securityGroup?: ec2.ISecurityGroup;
+  private networkConfiguration?: any;
+  private readonly integrationPattern: sfn.IntegrationPattern;
+
+  constructor(scope: cdk.Construct, id: string, private readonly props: EcsRunTaskProps) {
+    super(scope, id, props);
+    this.integrationPattern = props.integrationPattern || sfn.IntegrationPattern.REQUEST_RESPONSE;
+
+    validatePatternSupported(this.integrationPattern, EcsRunTask.SUPPORTED_INTEGRATION_PATTERNS);
+
+    if (this.integrationPattern === sfn.IntegrationPattern.WAIT_FOR_TASK_TOKEN && !sfn.FieldUtils.containsTaskToken(props.containerOverrides)) {
+      throw new Error('Task Token is required in `containerOverrides` for callback. Use Context.taskToken to set the token.');
+    }
+
+    this.validateLaunchTarget();
+
+    for (const override of this.props.containerOverrides || []) {
+      const name = override.containerDefinition.containerName;
+      if (!cdk.Token.isUnresolved(name)) {
+        const cont = this.props.taskDefinition.node.tryFindChild(name);
+        if (!cont) {
+          throw new Error(`Overrides mention container with name '${name}', but no such container in task definition`);
+        }
+      }
+    }
+
+    this.taskPolicies = this.makePolicyStatements();
+  }
+
+  protected renderTask(): any {
+    if (this.networkConfiguration !== undefined) {
+      // Make sure we have a security group if we're using AWSVPC networking
+      this.securityGroup = this.securityGroup ?? new ec2.SecurityGroup(this, 'SecurityGroup', { vpc: this.props.cluster.vpc });
+      this.connections.addSecurityGroup(this.securityGroup);
+    }
+
+    return {
+      Resource: integrationResourceArn('ecs', 'runTask', this.integrationPattern),
+      Parameters: sfn.FieldUtils.renderObject({
+        Cluster: this.props.cluster.clusterArn,
+        TaskDefinition: this.props.taskDefinition.taskDefinitionArn,
+        NetworkConfiguration: this.networkConfiguration,
+        Overrides: renderOverrides(this.props.containerOverrides),
+        ...this.props.launchTarget.bind(this).parameters,
+      }),
+    };
+  }
+
+  private configureAwsVpcNetworking() {
```

Comment:
nit: I like to order private (or any other) methods in the (approx) order in which they're called so it's easier to read and the reader need not jump up and down to follow the code path.

Reply:
sure, that sounds like a good guiding principle